src/Playwright/Core/Route.cs

```diff
@@ -69,15 +69,36 @@ public Task FulfillAsync(RouteFulfillOptions options = default)
                 options.BodyBytes,
                 options.Path);
 
-            return _channel.FulfillAsync(normalized);
+            return RaceWithPageCloseAsync(_channel.FulfillAsync(normalized));
         }
 
-        public Task AbortAsync(string errorCode = RequestAbortErrorCode.Failed) => _channel.AbortAsync(errorCode);
+        public Task AbortAsync(string errorCode = RequestAbortErrorCode.Failed) => RaceWithPageCloseAsync(_channel.AbortAsync(errorCode));
 
         public Task ContinueAsync(RouteContinueOptions options = default)
         {
             options ??= new RouteContinueOptions();
-            return _channel.ContinueAsync(url: options.Url, method: options.Method, postData: options.PostData, headers: options.Headers);
+            return RaceWithPageCloseAsync(_channel.ContinueAsync(url: options.Url, method: options.Method, postData: options.PostData, headers: options.Headers));
+        }
+
+        internal async Task InternalContinueAsync(RouteContinueOptions options = default)
+        {
+            try
+            {
+                await ContinueAsync(options).ConfigureAwait(false);
+            }
+            catch (Exception)
+            {
+            }
+        }
+
+        private async Task RaceWithPageCloseAsync(Task task)
+        {
+            var page = (Page)Request.Frame.Page;
+
+            // When page closes or crashes, we catch any potential rejects from this Route.
+            // Note that page could be missing when routing popup's initial request that
+            // does not have a Page initialized just yet.
+            await Task.WhenAny(task, page.ClosedOrCrashedTcs.Task).ConfigureAwait(false);
```

Comment:
`Task.WhenAny` is dangerous when not all tasks it takes are observed (e.g. accessing the `task.Exception` property when an exception occurs) which is the case here.
For example add the following code at the start of the `ShouldNotThrowWhenContinuingWhilePageIsClosing`, `ShouldNotThrowWhenContinuingAfterPageIsClosed` tests and set breakpoints in there:
```c#
TaskScheduler.UnobservedTaskException += (s, e) =>
{
    // set breakpoint here
};
```
Debug these tests `ShouldNotThrowWhenContinuingAfterPageIsClosed`, `ShouldNotThrowWhenContinuingWhilePageIsClosing` together and the `UnobservedTaskException` will be called because the `task.Exception`  property is never observed.

![image](https://user-images.githubusercontent.com/4009570/149840053-7b945f79-604b-4bad-8744-d9fadb323156.png)


I had this issue in my code as well and solved it like this, which will observe all `PlaywrightException`s:
```suggestion
            _ = task.ContinueWith(t => t.Exception.Handle(e => e is PlaywrightException), System.Threading.CancellationToken.None, TaskContinuationOptions.OnlyOnFaulted, TaskScheduler.Current);
            await Task.WhenAny(task, page.ClosedOrCrashedTcs.Task).ConfigureAwait(false);
```

Reply:
After looking at the code in more detail all the current callsites are already "racing" with the close event.
1. They might get an exception set here like `Target page, context or browser has been closed`:
https://github.com/microsoft/playwright-dotnet/blob/914c8fa6b3de7f5a62e76e059b10d18604064b01/src/Playwright/Transport/Connection.cs#L224-L227

2. And if not they will always get set an exception here:
https://github.com/microsoft/playwright-dotnet/blob/914c8fa6b3de7f5a62e76e059b10d18604064b01/src/Playwright/Transport/Connection.cs#L360-L363

Or is the intention to not throw when the page gets closed? If so then it needs to be suppress like I have shown above or this would also work:

```suggestion
            if (task != await Task.WhenAny(task, page.ClosedOrCrashedTcs.Task).ConfigureAwait(false))
            {
                try
                {
                    // this will resolve (almost) immediately as the exception is already set by one of the two methods mentioned above.
                    await task.ConfigureAwait(false); 
                }
                catch
                {
                    // ignore
                }
            }
```




On node the `process.on('unhandledRejection', function() {})` should also have been called I guess.

Reply:
What happens if we don't do it? Our assumption is that modern targets don't crash on unobserved exceptions.

Reply:
Yes, they don't crash anymore since `net45` but `TaskScheduler.UnobservedTaskException` is still raised, which might be surprising for users and likely create issues and they can't do anything about that.
https://docs.microsoft.com/dotnet/api/system.threading.tasks.taskscheduler.unobservedtaskexception?view=net-6.0

---

I ran into this the last week as I use `CancellationToken`s instead of timeouts so my code looked something like this:

```c#
var abortTask = Task.Delay(Timeout.Infinite, cancellationToken);
var page = await browserContext.NewPageAsync();
try
{
    var gotoTask = page.GotoAsync("...", new PageGotoOptions { Timeout = 0 });
    if (abortTask == await Task.WhenAny(gotoTask, abortTask))
    {
        cancellationToken.ThrowIfCancellationRequested();
    }
    await gotoTask;

    // ...
}
finally
{
    await page.CloseAsync();
}
```

Then I got `TaskScheduler.UnobservedTaskException` events because the `gotoTask` was unobserved when the cancellation token was canceled. And as we log all `UnobservedTaskException`s it created a lot of unnecessary noise.
So I added a extension method in which I ignored any `PlaywrightException`s with the message `Target closed`:

```c#
var abortTask = Task.Delay(Timeout.Infinite, cancellationToken);
var page = await browserContext.NewPageAsync();
try
{
    var gotoTask = page.GotoAsync("...", new PageGotoOptions { Timeout = 0 }).IgnorePlaywrightTargetClosedException();
    if (abortTask == await Task.WhenAny(gotoTask, abortTask))
    {
        cancellationToken.ThrowIfCancellationRequested();
    }
    await gotoTask;

    // ...
}
finally
{
    await page.CloseAsync();
}

static Task<T> IgnorePlaywrightTargetClosedException<T>(this Task<T> task)
{
    task.ContinueWith(t => t.Exception.Handle(e => e is PlaywrightException && e.Message.Contains("Target closed")));
    return task;
}
```

Reply:
@ campersau it would make sense to ensure in all our nunit tests that no such UnobservedTaskException expected are emitted. What do you think?

Reply:
Yes, that might be a good strategy to find `UnobservedTaskException`s in the product code. The problem is that in order to know which `UnobservedTaskException` is thrown in which test, they need to run in sequence instead of parallel, which  takes some time (13 minutes on my machine for one TFM).

Here is the code to check for `UnobservedTaskException` for each test: https://github.com/microsoft/playwright-dotnet/commit/3e29b6a062b6b7de2d1c025b233eb95a2cff6cf0
Here is a list of tests for which a `UnobservedTaskException` is currently thrown:
```
Fehler Microsoft:Playwright:Tests:BrowserContextBasicTests:CloseShouldAbortWaitForEvent [46 ms]
Fehler Microsoft:Playwright:Tests:BrowserContextCSPTests:ShouldBypassCSPHeader [1 s]
Fehler Microsoft:Playwright:Tests:BrowserContextCSPTests:ShouldBypassCSPInIframesAsWell [1 s]
Fehler Microsoft:Playwright:Tests:BrowserContextCSPTests:ShouldBypassCSPMetatag [1 s]
Fehler Microsoft:Playwright:Tests:BrowserTypeConnectTests:DisconnectedEventShouldBeEmittedWhenBrowserIsClosedOrServerIsClosed [2 s]
Fehler Microsoft:Playwright:Tests:BrowserTypeConnectTests:DisconnectedEventShouldHaveBrowserAsArguments [1 s]
Fehler Microsoft:Playwright:Tests:BrowserTypeConnectTests:ShouldBeAbleToConnectTwoBrowsersAtTheSameTime [1 s]
Fehler Microsoft:Playwright:Tests:BrowserTypeConnectTests:ShouldBeAbleToReconnectToABrowser [2 s]
Fehler Microsoft:Playwright:Tests:BrowserTypeConnectTests:ShouldEmitCloseEventsOnPagesAndContexts [2 s]
Fehler Microsoft:Playwright:Tests:BrowserTypeConnectTests:ShouldErrorWhenSavingDownloadAfterDeletion [2 s]
Fehler Microsoft:Playwright:Tests:BrowserTypeConnectTests:ShouldNotThrowOnCloseAfterDisconnect [2 s]
Fehler Microsoft:Playwright:Tests:BrowserTypeConnectTests:ShouldNotThrowOnContextCloseAfterDisconnect [2 s]
Fehler Microsoft:Playwright:Tests:BrowserTypeConnectTests:ShouldNotThrowOnPageCloseAfterDisconnect [1 s]
Fehler Microsoft:Playwright:Tests:BrowserTypeConnectTests:ShouldRejectNavigationWhenBrowserCloses [2 s]
Fehler Microsoft:Playwright:Tests:BrowserTypeConnectTests:ShouldRejectWaitForSelectorWhenBrowserCloses [2 s]
Fehler Microsoft:Playwright:Tests:BrowserTypeConnectTests:ShouldSaveAsVideosFromRemoteBrowser [3 s]
Fehler Microsoft:Playwright:Tests:BrowserTypeConnectTests:ShouldSaveDownload [2 s]
Fehler Microsoft:Playwright:Tests:BrowserTypeConnectTests:ShouldSaveHar [2 s]
Fehler Microsoft:Playwright:Tests:BrowserTypeConnectTests:ShouldSetTheBrowserConnectedState [1 s]
Fehler Microsoft:Playwright:Tests:BrowserTypeConnectTests:ShouldSupportSlowMo [2 s]
Fehler Microsoft:Playwright:Tests:BrowserTypeConnectTests:ShouldTerminateNetworkWaiters [1 s]
Fehler Microsoft:Playwright:Tests:BrowserTypeConnectTests:ShouldThrowWhenUsedAfterIsConnectedReturnsFalse [1 s]
Fehler Microsoft:Playwright:Tests:BrowserTypeConnectTests:ShouldThrowWhenWhenCallingWaitForNavigationAfterDisconnect [2 s]
Fehler Microsoft:Playwright:Tests:DefaultBrowserContext1Tests:ContextAddCookiesShouldWork [1 s]
Fehler Microsoft:Playwright:Tests:DefaultBrowserContext1Tests:ContextClearCookiesShouldWork [1 s]
Fehler Microsoft:Playwright:Tests:DefaultBrowserContext1Tests:ContextCookiesShouldWork [1 s]
Fehler Microsoft:Playwright:Tests:DefaultBrowserContext1Tests:ShouldNotBlockThirdPartyCookies [3 s]
Fehler Microsoft:Playwright:Tests:DefaultBrowserContext1Tests:ShouldSupportBypassCSPOption [1 s]
Fehler Microsoft:Playwright:Tests:DefaultBrowserContext1Tests:ShouldSupportHttpCredentialsOption [1 s]
Fehler Microsoft:Playwright:Tests:DefaultBrowserContext1Tests:ShouldSupportOfflineOption [895 ms]
Fehler Microsoft:Playwright:Tests:DefaultBrowserContext1Tests:ShouldSupportUserAgentOption [1 s]
Fehler Microsoft:Playwright:Tests:DefaultBrowsercontext2Tests:ShouldSupportExtraHTTPHeadersOption [1 s]
Fehler Microsoft:Playwright:Tests:DefaultBrowsercontext2Tests:ShouldSupportGeolocationAndPermissionsOptions [1 s]
Fehler Microsoft:Playwright:Tests:DefaultBrowsercontext2Tests:ShouldSupportHasTouchOption [1 s]
Fehler Microsoft:Playwright:Tests:DefaultBrowsercontext2Tests:ShouldSupportIgnoreHTTPSErrorsOption [1 s]
Fehler Microsoft:Playwright:Tests:DefaultBrowsercontext2Tests:ShouldWorkInPersistentContext [1 s]
Fehler Microsoft:Playwright:Tests:DownloadTests:ShouldExposeStream [1 s]
Fehler Microsoft:Playwright:Tests:PageBasicTests:ShouldFailWithErrorUponDisconnect [183 ms]
Fehler Microsoft:Playwright:Tests:PageBasicTests:ShouldTerminateNetworkWaiters [286 ms]
Fehler Microsoft:Playwright:Tests:PageClickTimeout2Tests:ShouldTimeoutWaitingForVisibilityHiddenToBeGone [5 s]
Fehler Microsoft:Playwright:Tests:PageGotoTests:ShouldNotCrashWhenNavigatingToBadSSLAfterACrossOriginNavigation [495 ms]
Fehler Microsoft:Playwright:Tests:PageRequestContinueTests:ShouldNotThrowWhenContinuingAfterPageIsClosed [175 ms]
Fehler Microsoft:Playwright:Tests:PageRouteTests:ShouldBeAbortableWithCustomErrorCodes [167 ms]
Fehler Microsoft:Playwright:Tests:PopupTests:ShouldRespectRoutesFromBrowserContext [709 ms]
Fehler Microsoft:Playwright:Tests:PopupTests:ShouldRespectRoutesFromBrowserContextUsingWindowOpen [545 ms]
```

Complete output (german): [output.txt](https://github.com/microsoft/playwright-dotnet/files/7895483/output.txt)

Resources:
- https://github.com/nunit/nunit/issues/2220#issue-233750693
- https://github.com/nunit/nunit/issues/2220#issuecomment-430106577
- https://stackoverflow.com/a/3284286

---

For example with the `Microsoft:Playwright:Tests:BrowserTypeConnectTests:ShouldSaveDownload` test I discovered that this code is missing an `await` as it currently raises a `UnobservedTaskException`:
https://github.com/microsoft/playwright-dotnet/blob/829e0d41804809f85f683063623fb3d437d48cb7/src/Playwright/Core/BrowserType.cs#L150

FYI: just discarding the task with `_ = ...` does not work as the `task.Exception` is still unobserved!

Reply:
> FYI: just discarding the task with _ = ... does not work as the task.Exception is still unobserved!

yeah this is known. We can get in theory rid of all `_ = ...` things and just keep the `...` afaik.

I'd like to follow-up on that then to keep the diff small of the PR, I rebased this PR back to its original form. We can then add an additional GitHub Action job which runs the tests only in chromium and verifies that there are no unhandled exceptions thrown. This might be better instead of increasing all the jobs up to 13+ minutes.

I started locally with adding a method onto the Task calling it FireAndForget() which is like `myAction().catch(() => {})` in JS. If you are interested in adding the GitHub Action job, let me know, otherwise I can do it too. Thanks!

Reply:
Okay! Yes, I can add the GitHub Action job for that.

Reply:
Let's annotate it with _ = for now.

Reply:
Just noticed that a new GitHub Action job isn't needed since the tests are already running in sequence in CI. `NUnit.NumberOfTestWorkers=1`

So here are all the changes required to make tests pass without raising `UnobservedTaskException`:
https://github.com/microsoft/playwright-dotnet/compare/main...campersau:UnobservedTaskExceptions
Here is a sample CI run: https://github.com/campersau/playwright-dotnet/actions/runs/1718338242

But I will wait with the PR for your `FireAndForget` and adjust to it accordingly.